internal/services/gitserver/repo-cleaner.go

```diff
@@ -0,0 +1,153 @@
+package gitserver
+
+import (
+	"context"
+	"io/ioutil"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"agola.io/agola/internal/util"
+)
+
+func (s *Gitserver) repoCleanerLoop(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("repoCleaner exiting")
+
+			return
+		case <-time.After(s.c.ScanTime):
+			if err := s.scanRepos(ctx); err != nil {
+				log.Errorf("err: %+v", err)
+			}
+		}
+	}
+}
+
+func (s *Gitserver) scanRepos(ctx context.Context) error {
+	log.Info("repoCleaner scanRepos start")
+
+	usersDir, err := ioutil.ReadDir(s.c.DataDir)
+	if err != nil {
+		return err
+	}
+
+	for _, u := range usersDir {
+		if u.IsDir() {
+			reposDir, _ := ioutil.ReadDir(filepath.Join(s.c.DataDir, u.Name()))
+			for _, r := range reposDir {
+				if r.IsDir() {
+					s.scanRepo(ctx, filepath.Join(s.c.DataDir, u.Name(), r.Name()))
+				}
+			}
+		}
+	}
+
+	log.Info("repoCleaner scanRepos end")
+
+	return nil
+}
+
+func (s *Gitserver) scanRepo(ctx context.Context, repoDir string) {
+	git := &util.Git{GitDir: repoDir}
+
+	branches, _ := s.getBranches(git, ctx)
+	deletedCount := 0
+	for _, b := range branches {
+		date, err := s.getLastCommitTime(ctx, git, "refs/heads/"+b)
+		if err != nil {
+			log.Error("repoCleaner error:", err)
+			continue
+		}
+
+		if time.Since(date) >= s.c.RepositoryRefsExpireInterval {
+			err := s.deleteBranch(ctx, git, b)
+			if err != nil {
+				log.Error("repoCleaner error: %v", err)
+			} else {
+				deletedCount++
+			}
+		}
+	}
+
+	tags, _ := s.getTags(git, ctx)
+	for _, tag := range tags {
+		date, err := s.getLastCommitTime(ctx, git, "refs/tags/"+tag)
+		if err != nil {
+			log.Error("repoCleaner error:", err)
+			continue
+		}
+
+		if time.Since(date) >= s.c.RepositoryRefsExpireInterval {
+			err := s.deleteTag(ctx, git, tag)
+			if err != nil {
+				log.Error("repoCleaner error:", err)
+			} else {
+				deletedCount++
+			}
+		}
+	}
+
+	if len(branches)+len(tags) == deletedCount {
+		log.Info("deleting repo:", repoDir)
+		err := s.deleteRepo(ctx, repoDir)
+		if err != nil {
+			log.Error("repoCleaner error:", err)
+		}
+	}
+}
+
+func (s *Gitserver) getBranches(git *util.Git, ctx context.Context) ([]string, error) {
+	branches, err := git.OutputLines(ctx, nil, "for-each-ref", "--format=%(refname:short)", "refs/heads/")
+	if err != nil {
+		return nil, err
+	}
+
+	return branches, nil
+}
+
+func (s *Gitserver) getTags(git *util.Git, ctx context.Context) ([]string, error) {
+	tags, err := git.OutputLines(ctx, nil, "for-each-ref", "--format=%(refname:short)", "refs/tags/")
+	if err != nil {
+		return nil, err
+	}
+
+	return tags, nil
+}
+
+func (s *Gitserver) getLastCommitTime(ctx context.Context, git *util.Git, ref string) (time.Time, error) {
+	output, err := git.Output(ctx, nil, "log", "-1", "--format=%cd", ref)
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	date, err := time.Parse("Mon Jan 2 15:04:05 2006 -0700", strings.TrimSuffix(string(output), "\n"))
```

Comment:
git log --format=%cd is env dependant. What about using `%cI` (Strict ISO 8601) and then use `time.Parse(time.RF3339, ....)`?